nexus/src/internal_api/http_entrypoints.rs

```diff
@@ -51,1060 +68,779 @@ use omicron_common::api::internal::nexus::RepairProgress;
-) -> Result<HttpResponseUpdatedNoContent, HttpError> {
-    let apictx = rqctx.context();
-    let nexus = &apictx.nexus;
-    let opctx = crate::context::op_context_for_internal_api(&rqctx).await;
-    let path = path_params.into_inner();
-    let info = sled_info.into_inner();
-    let sled_id = &path.sled_id;
-    let handler = async {
-        nexus.upsert_sled(&opctx, *sled_id, info).await?;
-        Ok(HttpResponseUpdatedNoContent())
-    };
-    apictx.internal_latencies.instrument_dropshot_handler(&rqctx, handler).await
-}
-
-/// Request a new set of firewall rules for a sled.
-///
-/// This causes Nexus to read the latest set of rules for the sled,
-/// and call a Sled endpoint which applies the rules to all OPTE ports
-/// that happen to exist.
-#[endpoint {
-     method = POST,
-     path = "/sled-agents/{sled_id}/firewall-rules-update",
- }]
-async fn sled_firewall_rules_request(
-    rqctx: RequestContext<Arc<ServerContext>>,
-    path_params: Path<SledAgentPathParam>,
-) -> Result<HttpResponseUpdatedNoContent, HttpError> {
-    let apictx = rqctx.context();
-    let nexus = &apictx.nexus;
-    let opctx = crate::context::op_context_for_internal_api(&rqctx).await;
-    let path = path_params.into_inner();
-    let sled_id = &path.sled_id;
-    let handler = async {
-        nexus.sled_request_firewall_rules(&opctx, *sled_id).await?;
-        Ok(HttpResponseUpdatedNoContent())
-    };
-    apictx.internal_latencies.instrument_dropshot_handler(&rqctx, handler).await
+pub(crate) struct NexusInternalApiImpl {
+    pub(crate) context: Arc<ServerContext>,
 }
 
-/// Path parameters for Rack requests.
-#[derive(Deserialize, JsonSchema)]
-struct RackPathParam {
-    rack_id: Uuid,
-}
-
-/// Report that the Rack Setup Service initialization is complete
-///
-/// See RFD 278 for more details.
-#[endpoint {
-     method = PUT,
-     path = "/racks/{rack_id}/initialization-complete",
- }]
-async fn rack_initialization_complete(
-    rqctx: RequestContext<Arc<ServerContext>>,
-    path_params: Path<RackPathParam>,
-    info: TypedBody<RackInitializationRequest>,
-) -> Result<HttpResponseUpdatedNoContent, HttpError> {
-    let apictx = rqctx.context();
-    let nexus = &apictx.nexus;
-    let path = path_params.into_inner();
-    let request = info.into_inner();
-    let opctx = crate::context::op_context_for_internal_api(&rqctx).await;
-
-    nexus.rack_initialize(&opctx, path.rack_id, request).await?;
-
-    Ok(HttpResponseUpdatedNoContent())
-}
-
-/// Path parameters for Switch requests.
-#[derive(Deserialize, JsonSchema)]
-struct SwitchPathParam {
-    switch_id: Uuid,
-}
-
-#[endpoint {
-    method = PUT,
-    path = "/switch/{switch_id}",
-}]
-async fn switch_put(
-    rqctx: RequestContext<Arc<ServerContext>>,
-    path_params: Path<SwitchPathParam>,
-    body: TypedBody<SwitchPutRequest>,
-) -> Result<HttpResponseOk<SwitchPutResponse>, HttpError> {
-    let apictx = rqctx.context();
-    let handler = async {
+#[async_trait::async_trait]
```

Comment:
For now I've chosen to use `async_trait`. But note that we don't need dynamic dispatch, so it should be possible to use the `-> impl Future` support in current versions of Rust.

Main thing to watch out for, however, is that we'd have to add `Send` bounds to all the returned Futures in the trait. That would change the types to `impl Future<...> + Send`, which would no longer make it possible to just write `async fn`. (Hopefully this is addressed soon.)

Reply:
Removed async-trait -- works great!